### PR TITLE
Updates to log viewer for 4.3

### PR DESCRIPTION
**Pull request by ctzsnooze**
Adds log header support for the new dynamic notch, rc_smoothing, and feedforward fields in 4.3 after inclusion of firmware PR #10723

A 4.2 or earlier log will display values in the appropriate boxes where possible, or in the 'unknown fields' box at the bottom.

This shows how the feedforward part of a log header appears with a 4.3 log

Screen Shot 2021-07-05 at 07 31 27

And here how the dynamic notch values look for 4.3

Screen Shot 2021-07-05 at 07 31 55

And rcSmoothing:

Screen Shot 2021-07-05 at 07 30 31

This supersedes the individual PRs #510 and #511.

### CODE REVIEW

**Comment by McGiverGim** [comment]
I'm not too sure what you want to do: you have "removed" some code that lets old versions show the values in a correct manner, and only the new one is shown ok. Is difficult to me to follow the changes with all of this code removed...

We must get to a solution where 4.2 logs are shown, at least, as good as in the current blackbox version, and 4.3 logs are shown in a correct way.

The first to know, to fix that, is what are the differences between versions. Are the fields simply "renamed" or some of them have different values? I see some differences in the enumerations. Can you explain it here? I think I can point you to the correct changes then.

If some values have simply being renamed, you can "mark" this in the Blackbox here: https://github.com/betaflight/blackbox-log-viewer/blob/7da5b001e1253c470e0e9670138c5a3c162422d6/js/flightlog_parser.js#L316
In this way we "rename" some values from the header to the internal var, so you can make two different headers behave exactly the same.

And about the fielddefs definition, we have a method at the end where we "adapt" them to old/new versions:
https://github.com/betaflight/blackbox-log-viewer/blob/7da5b001e1253c470e0e9670138c5a3c162422d6/js/flightlog_fielddefs.js#L462

Now I think you understand better the nightmare of backwards compatibility, when you work only in the firmware part you are unaware of all of that...

**Comment by klutvott123** [comment]
I think we should wait with the dynamic notch changes. @KarateBrot says we're going back to using Q instead of bandwidth as there weren't any benefits to using constant bandwidth.
@KarateBrot What's the status on this?

**Comment by KarateBrot** [comment]
@klutvott123 Yes, thanks for your objection. 

==> **Bandwidth can be ignored in this PR** <==

I need to get the dynamic notch PR ready asap (probably tomorrow) to inform everyone.
I talked with @Asizon and it looks like it's no big deal to revert the change for the configurator as well.

**Comment by ctzsnooze** [comment]
@McGiverGim   Thanks for your advice! I have tried to do the right things and apologise if it is not done right.  I checked that a 4.3 and a 4.2 log would open values in the right places, that is the limit of my testing.

The following fields are added in this PR.  They are mostly new from 4.3, but some existed before and have not been added previously.
```
ff_transition
ff_averaging
ff_smooth_factor
ff_jitter_factor
ff_boost
ff_max_rate_limit

rc_smoothing_mode
rc_smoothing_feedforward_hz
rc_smoothing_setpoint_hz
rc_smoothing_auto_factor_setpoint
rc_smoothing_throttle_hz
rc_smoothing_auto_factor_throttle
rc_smoothing_active_cutoffs_ff_sp_thr

rpm_notch_lpf

dyn_notch_count
```

These fields are not used in 4.3, and have no specific translation as they aren't really updating a previous version.
```
rc_smoothing_auto_factor  (now two separate factors, one for throttle, one for set point)
rc_smoothing_type (not used at all in 4.3)
rc_smoothing_filter_type (not used at all in 4.3)
dyn_notch_width_percent (not used at all in 4.3.  We now have dyn_notch_count in a kind of analogous role)
```

The `adjustFieldDefsList` concept seems really clever, but I don't know if it should be applied to any of these changes, and I wouldn't feel confident changing it.

I've removed dyn_notch_bandwidth from the original proposal now we are going to stay with dyn_notch_q.

I attempted to resolve the Soundcloud 'bugs'.

Please let me know what else I need to do.

**Comment by McGiverGim** [comment]
It seems something is wrong with this PR, the changes have been lost...
I will try to help you when you have the code up again.

File: js/flightlog_parser.js
Comment (McGiverGim) [inline js/flightlog_parser.js:313]: Why we have two variables that are the same? This `ff_transition`and the one of line 223 `feedforward_transition`?

I think you are understanding bad how this works: you must have only **one** variable to store the value. With the name you want, and use the translation table to "translate" the content of the blackbox header to the internal name if it is not the same. In this way, if this parameter has different names between versions, we will have only one internal variable, and all the translations needed in the translations table to assign it correctly.

File: js/flightlog_fielddefs.js
Comment (McGiverGim) [inline js/flightlog_fielddefs.js:236]: Remove the OFF option will let old versions that used it without backward compatibility. What we do here is to have the "basic" enumeration, and modify it for some versions below, in the method of the line 467.

The same applies to the FILTER_TYPE.

**Review by McGiverGim — CHANGES_REQUESTED** [review]
The base is wrong. A first review asking for changes in the basic, once we have this right, we can go for the rest.

File: js/flightlog_parser.js
Comment (ctzsnooze) [reply js/flightlog_parser.js:313]: It is like that because I do not know how to do it correctly.  I could not find an example in the translation table to follow.  And this worked, even if it was bad  :-)

If you could me exactly what I must do to fix this particular example, then I can do it correctly for the others.

File: js/flightlog_fielddefs.js
Comment (ctzsnooze) [reply js/flightlog_fielddefs.js:236]: Yes however again I could not figure out the correct way to do that.  

Again, if you could show exactly what I must do for one of them, I will do for the others that are similar.  I just couldn't see an example to follow.

File: js/flightlog_parser.js
Comment (McGiverGim) [reply js/flightlog_parser.js:313]: This part is easy, I suppose you want to use the "new" name (if not, do the inverse of what I say here). The process is like this:
1. Remove the old declaration: 
```javascript
feedforward_transition:null,            // Feedforward transition
```
2. Add the new declaration:
```javascript
ff_transition:null,            // Feedforward transition
```
2. Add the value to the `translationValues` table, to maintain backwards compatibility:
```javascript
feedforward_transition : "ff_transition",
``` 
3. Replace all the references to the old `feedforward_transition` to the new `ff_transition` in the code.

File: js/flightlog_fielddefs.js
Comment (McGiverGim) [reply js/flightlog_fielddefs.js:236]: There are several ways, for example, for RC_SMOOTHING_DERIVATIVE_TYPE (I do this without computer, maybe something is wrong):
1. Change the actual declaration for this:
```javascript
RC_SMOOTHING_DERIVATIVE_TYPE = [],
RC_SMOOTHING_DERIVATIVE_TYPE_COMPLETE = makeReadOnly([
    "OFF",
    "PT1",
    "BIQUAD",
]),
```
2. In the `adjustFieldDefsList` function, there are different ways to add it, for example something like:
```javascript
if (semver.gte(firmwareVersion, '4.3.0')) {
    RC_SMOOTHING_DERIVATIVE_TYPE = RC_SMOOTHING_DERIVATIVE_TYPE_COMPLETE.slice(0);
    RC_SMOOTHING_DERIVATIVE_TYPE.splice(RC_SMOOTHING_DERIVATIVE_TYPE .indexOf('OFF'), 1);
    RC_SMOOTHING_DERIVATIVE_TYPE = makeReadOnly(RC_SMOOTHING_DERIVATIVE_TYPE);
} else {
    RC_SMOOTHING_DERIVATIVE_TYPE = makeReadOnly(RC_SMOOTHING_DERIVATIVE_TYPE_COMPLETE);
}
```

File: js/flightlog_parser.js
Comment (ctzsnooze) [reply js/flightlog_parser.js:312]: I have a working version of the log viewer where I can edit directly the .js files.
If I attempt these changes it will not open a log file. 
I am doing something wrong.  
I'll push a commit, but it seems to not work properly.

File: js/flightlog_parser.js
Comment (McGiverGim) [inline js/flightlog_parser.js:774]: Here happens the same than with `ff_transition`: two variables that are the same. You must "translate" one into the other and let only one.
The only difference, I suppose that by error, we don't have a "default" value in the original code for it.

File: js/flightlog_parser.js
Comment (McGiverGim) [inline js/flightlog_parser.js:1553]: This is from other PR... this don't need to be here. A rebase of your code will include it by default, it was merged.

File: js/flightlog_parser.js
Comment (McGiverGim) [reply js/flightlog_parser.js:312]: I've not tested it but seems ok :)

**Review by McGiverGim — DISMISSED** [review]

File: js/flightlog_parser.js
Comment (ctzsnooze) [reply js/flightlog_parser.js:312]: I tested, works great!  Thanks for explaining how to do it properly.

File: js/flightlog_fielddefs.js
Comment (ctzsnooze) [reply js/flightlog_fielddefs.js]: This is too difficult for me :-)  If definitely it needs to be done, I'll have to leave that 'fix' to someone with your level of understanding.
Most of the old enums can remain unchanged, I hope.
For the `FILTER_TYPE` enum, I'm hoping it's OK to remove the FIR filter, which (as far as I know) hasn't been used for many years, and added the extra PT2 and PT3 members on the end.  I hope this is OK.

File: js/flightlog_parser.js
Comment (ctzsnooze) [reply js/flightlog_parser.js:774]: Fixed in the same way as above, appears to work correctly on testing with 4.3 and 4.2

File: js/flightlog_parser.js
Comment (ctzsnooze) [reply js/flightlog_parser.js:1553]: Yes I'll fix that

File: index.html
Comment (Asizon) [inline index.html:1279]: Here is max of 255 for Notch Q, thats correct @ctzsnooze ?

File: js/flightlog_fielddefs.js
Comment (Asizon) [inline js/flightlog_fielddefs.js:256]: we are removing backward compatiblity for it?

File: index.html
Comment (KarateBrot) [reply index.html:1279]: Right now `dyn_notch_q` can go up to 1000 (min=1, max=1000).

Comment (ctzsnooze) [reply index.html:1279]: @McGiverGim I think it's OK as is, but really haven't a clue.  Please let me know if there's anything that needs changing.

File: js/flightlog_fielddefs.js
Comment (ctzsnooze) [reply js/flightlog_fielddefs.js:256]: FIR was never implemented in firmware, and I'm only adding PT2 and PT3 to the enum... 
I don't know if this is OK or not, at this point, and am awaiting further advice.

Comment (McGiverGim) [reply js/flightlog_fielddefs.js:256]: It does not depend if it was implemented or not... It depends on the value the firmware wrote to the log, as int, this enum must be consistent

Comment (ctzsnooze) [reply js/flightlog_fielddefs.js:256]: What I mean is that we have had PT1 and Biquad as the only options before.  So 0 was always PT1 and 1 always Biquad.   The only change now is to add PT2 and PT3 on the end as options 2 and 3.  These were not used previously, as far as I am aware.

Comment (McGiverGim) [reply js/flightlog_fielddefs.js:256]: We can decide if version 3.3.3 needs to be supported or not. It seems it existed, and SLEW too:

https://github.com/betaflight/betaflight/blob/3565182d311e5151526f8bb6c2d8f3c9192cd735/src/main/common/filter.h#L65-L70

Comment (ctzsnooze) [reply js/flightlog_fielddefs.js:256]: My personal feeling is that we should not worry about 3.3.3, since it was:
- 'unusual' in its filter options, they were 'experimental', I think;
- FIR was rarely, if ever, used, in practice;
- it was a long time ago, and
- 3.5.7 is the '3' version that people fly nowadays, it has PT1 and biquad options only.
Since I don't know how to code enum version control, I would be grateful if I could leave the PR as is.
If you know a good way to provide version control over enums, and if you had the time to implement a PR that would provide this, I'm sure there will be occasions in future where we may find that useful, so I leave that for you to decide.

File: js/header_dialog.js
Comment (Asizon) [inline js/header_dialog.js:119]: This parameter doesnt exist.

File: index.html
Comment (Asizon) [reply index.html:1279]: @ctzsnooze here is max value of 255 for Q when it should be 999.0 thought.

**Review by Asizon — CHANGES_REQUESTED** [review]

File: index.html
Comment (haslinghuis) [inline index.html:1272]: indentation

File: js/header_dialog.js
Comment (haslinghuis) [inline js/header_dialog.js:634]:
```suggestion
        if (activeSysConfig.firmwareType === FIRMWARE_TYPE_BETAFLIGHT && semver.gte(activeSysConfig.firmwareVersion, '4.3.0')) {
```

File: js/header_dialog.js
Comment (haslinghuis) [inline js/header_dialog.js:657]:
```suggestion
        if (activeSysConfig.firmwareType === FIRMWARE_TYPE_BETAFLIGHT && semver.gte(activeSysConfig.firmwareVersion, '4.3.0')) {
```

File: js/header_dialog.js
Comment (haslinghuis) [inline js/header_dialog.js:710]:
```suggestion
        if (activeSysConfig.firmwareType === FIRMWARE_TYPE_BETAFLIGHT && semver.gte(activeSysConfig.firmwareVersion, '4.3.0')) {
```

File: js/header_dialog.js
Comment (haslinghuis) [inline js/header_dialog.js:712]:
```suggestion
            setParameter('feedforwardSmoothing'  ,sysConfig.ff_smooth_factor, 0);
```

File: js/header_dialog.js
Comment (haslinghuis) [inline js/header_dialog.js:713]:
```suggestion
            setParameter('feedforwardJitter'     ,sysConfig.ff_jitter_factor, 0);
```

File: js/header_dialog.js
Comment (haslinghuis) [inline js/header_dialog.js:714]:
```suggestion
            setParameter('feedforwardMaxRate'    ,sysConfig.ff_max_rate_limit, 0);
```

File: js/header_dialog.js
Comment (haslinghuis) [inline js/header_dialog.js:716]: Same on next 5 lines

**Review by haslinghuis — CHANGES_REQUESTED** [review]
Commenting on code quality only

File: index.html
Comment (ctzsnooze) [reply index.html]: fixed
also replaced all tabs with four spaces

**Comment by ctzsnooze** [comment]
@haslinghuis - thanks for your suggestions and checks.  I've done those whitespace tidy ups.
I'd really be grateful if this could be merged soon, it fixes issues for 4.2 users and is helpful for all 4.3 developers.

**Comment by ctzsnooze** [comment]
Checked and tested on 4.3 and 4.2 logs.  Works well.  Should also be OK for rcSmoothing with earlier logs.
Happy to squash before merging.
Ready to merge, I think.

**Review by haslinghuis — DISMISSED** [review]

File: js/flightlog_fielddefs.js
Comment (mikeller) [reply js/flightlog_fielddefs.js:256]: While I agree with the statement that we should not worry about 3.3.3 when it comes to displaying everything correctly, I think there are some things to be considered:
- I think while people are not necessarily flying 3.3.3 any more, they still might have old logs from 3.3.3 around, and might want to open them with blackbox explorer.
- in this case displaying wrong data without at least a warning is misleading
What are thoughts around either adding a warning when opening old logs, telling the user that some data may be incorrect?

Comment (ctzsnooze) [reply js/flightlog_fielddefs.js:256]: Good point.  Maybe if a version of the log viewer doesn't fully logs before some firmware version, then we can show a message recommending to use an earlier version of the log viewer.  The message could be 'smart' and say which fields may not be displayed correctly, if we know.
Currently there is a lot of maintenance work required in the log viewer.

File: js/header_dialog.js
Comment (haslinghuis) [inline js/header_dialog.js:97]: 3.5 or 4.3 ?

Comment (ctzsnooze) [reply js/header_dialog.js:97]: Transition first came in back in 3.5.  It was called D weight back then, from memory.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by ctzsnooze** [comment]
Thanks @blckmn this should be really good!

**Comment by haslinghuis** [comment]
@ctzsnooze first conditions fails as one of the others fail. We could suppress output in this case but helps identify what's holding the auto-merge process.

**Review by haslinghuis — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Review by mikeller — COMMENTED** [review]
@McGiverGim, @Asizon: Are you ok with this, or are there still problems with backwards compatibility?

**Review by Asizon — DISMISSED** [review]
Looks really better now @mikeller

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by ctzsnooze** [comment]
rebased and tested

**Comment by ctzsnooze** [comment]
This is how the info screen looks for a 4.3 log:

Screen Shot 2021-09-03 at 21 42 28

File: index.html
Comment (McGiverGim) [inline index.html:926]: Why the transition has been removed? Betaflight 4.2 uses it if I'm not wrong...

Comment (ctzsnooze) [reply index.html:926]: I think `ptermSRateWeight` is removed, but transition is still there, and still shows up normally when looking at 4.2 logs.

**Review by haslinghuis — APPROVED** [review]

**Review by blckmn — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by mikeller — APPROVED** [review]

**Pull request merged**
